src/Workspaces/Remote/Core/RemoteCallback.cs

```diff
@@ -87,23 +88,106 @@ public async ValueTask<TResult> InvokeAsync<TResult>(Func<T, CancellationToken,
+        /// <param name="invocation">A callback to asynchronously write data. The callback should always <see
+        /// cref="PipeWriter.Complete"/> the <see cref="PipeWriter"/>.  If it does not then reading will hang</param>
+        /// <param name="reader">A callback to asynchronously read data. The callback should not complete the <see
+        /// cref="PipeReader"/>, but no harm will happen if it does.</param>
+        /// <param name="cancellationToken">A cancellation token the operation will observe.</param>
         public async ValueTask<TResult> InvokeAsync<TResult>(
             Func<T, PipeWriter, CancellationToken, ValueTask> invocation,
             Func<PipeReader, CancellationToken, ValueTask<TResult>> reader,
             CancellationToken cancellationToken)
         {
             try
             {
-                return await BrokeredServiceConnection<T>.InvokeStreamingServiceAsync(_callback, invocation, reader, cancellationToken).ConfigureAwait(false);
+                cancellationToken.ThrowIfCancellationRequested();
+                var pipe = new Pipe();
+
+                // Kick off the work to do the writing to the pipe asynchronously.  It will start hot and will be able
+                // to do work as the reading side attempts to pull in the data it is writing.
+
+                var writeTask = WriteAsync(_callback, pipe.Writer);
+                var readTask = ReadAsync(pipe.Reader);
+
+                // Note: waiting on the write-task is not strictly necessary.  The read-task cannot complete unless it
+                // the write-task completes (or it faults for some reason).  However, it's nice and clean to just not
+                // use fire-and-forget here and avoids us having to consider things like async-tracking-tokens for
+                // testing purposes.
+                await Task.WhenAll(writeTask, readTask).ConfigureAwait(false);
+                return await readTask.ConfigureAwait(false);
             }
             catch (Exception exception) when (ReportUnexpectedException(exception, cancellationToken))
             {
                 throw new OperationCanceledIgnoringCallerTokenException(exception);
             }
+
+            async Task WriteAsync(T service, PipeWriter pipeWriter)
+            {
+                try
+                {
+                    // Intentionally yield this thread so that the caller can proceed concurrently and start reading.
+                    await TaskScheduler.Default;
+
+                    await invocation(service, pipeWriter, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    // Ensure that the writer is complete if an exception is thrown. This intentionally swallows the
+                    // exception on this side, knowing it will actually be thrown on the reading side.
```

Comment:
> knowing it will actually be thrown on the reading side.

Strictly speaking, it will only throw on the other side if the writer hasn't _already_ been completed. It's presumably possible for your `invocation` delegate to throw *after* completing the writer, and that exception would be dropped by your next statement because you can only complete the writer once (subsequent calls are no-ops).

Reply:
good point.

Reply:
do you think we should add an explicit `throw;` here to be safe against that?